File: LibraryAndTools/UnionFind.cpp
class UnionFind {
private:
    vector<int> parent;
    vector<int> componentSize;
    int n, numberOfDisjointSets;
public:
    UnionFind(int _n) {
        n = _n;
        for (int i = 0; i < n; i++) {
            parent.push_back(i);
            componentSize.push_back(1);
        }
        numberOfDisjointSets = n;
    }

    int findParent(int x) {
        if (parent[x] == x)
            return x;
        return findParent(parent[x]);
    }

    void unionSets(int x, int y) {
        int root0 = findParent(x);
        int root1 = findParent(y);

        if (root0 == root1) return;

        if (componentSize[root0] >= componentSize[root1]) {
            componentSize[root0] += componentSize[root1];
            parent[root1] = root0;
        } else {
            componentSize[root1] += componentSize[root0];
            parent[root0] = root1;
        }
        numberOfDisjointSets--;
    }

    bool isSameComponent(int x, int y) {
        return (findParent(x) == findParent(y));
    }

    int getNumberOfDisjointSets() {
        return numberOfDisjointSets;
    }
};

File: LibraryAndTools/union_find.py
class SetUnion:
    def __init__(self, n):
        self.__n = n
        self.__parent, self.__size = [], []
        for j in range(n):
            self.__parent.append(j)
            self.__size.append(0)

    def find(self, x):
        return x if self.__parent[x] == x else self.find(self.__parent[x])

    def union_sets(self, x, y):
        root1, root2 = self.find(x), self.find(y)
        if root1 == root2:
            return
        if self.__size[root1] >= self.__size[root2]:
            self.__size[root1] += self.__size[root2]
            self.__parent[root2] = root1
        else:
            self.__size[root2] += self.__size[root1]
            self.__parent[root1] = root2

    def is_same_component(self, x, y):
        return self.find(x) == self.find(y)